src/parser.ts

```diff
@@ -1,4 +1,163 @@
-import { ValidationError } from "./common";
+/**
+ * The parser is responsible for taking a raw JS object
+ * of the XTP Schema and parsing and typing it without transforming
+ * any of the values. It's just mean to give a typed representation
+ * of the raw format.
+ */
+import { ValidationError } from "./common"
+
+export interface ParseResult {
+  doc?: VUnknownSchema;
+  errors?: ValidationError[];
+}
+
+/**
+ * Parses and validates an untyped object into a V*Schema
+ */
+export function parseAny(doc: any): ParseResult {
+  switch (doc.version) {
+    case 'v0':
+      return { doc: doc as V0Schema }
+    case 'v1-draft':
+      const v1Doc = doc as V1Schema
+      const validator = new V1Validator(v1Doc)
+      const errors = validator.validate()
+      return { doc: v1Doc, errors }
+    default:
+      return {
+        errors: [
+          new ValidationError(`version property not valid: ${doc.version}`, "#/version")
+        ]
+      }
+  }
+}
+
+export function isV0Schema(schema?: VUnknownSchema): schema is V0Schema {
+  return schema?.version === 'v0';
+}
+
+export function isV1Schema(schema?: VUnknownSchema): schema is V1Schema {
+  return schema?.version === 'v1-draft';
+}
+
+/**
+ * Validates a V1 document.
+ */
+class V1Validator {
+  errors: ValidationError[]
+  location: string[]
+  doc: any
+
+  constructor(doc: V1Schema) {
+    this.doc = doc as any
+    this.errors = []
+    this.location = ['#']
+  }
+
+  /**
+   * Validate the document and return any errors
+   */
+  validate(): ValidationError[] {
+    this.errors = []
+    this.validateNode(this.doc)
+    return this.errors
+  }
+
+  /**
+   * Recursively walk through the untyped document and validate each node.
+   * This saves us a lot of code but might be a little bit slower.
+   */
+  validateNode(node: any) {
+    this.validateTypedInterface(node)
+
+    if (node && typeof node === 'object') {
+      // i don't think we need to validate array children
+      if (Array.isArray(node)) return
+
+      for (const key in node) {
+        if (Object.prototype.hasOwnProperty.call(node, key)) {
+          const child = node[key]
+          if (typeof child === 'object') {
+            this.location.push(key)
+            this.validateNode(child);
+            this.location.pop()
+          }
+        }
+      }
+    }
+  }
+
+  recordError(msg: string) {
+    this.errors.push(
+      new ValidationError(msg, this.getLocation())
+    )
+  }
+
+  /**
+   * Validates that a node conforms to the rules of
+   * the XtpTyped interface. Validates what we can't
+   * catch in JSON Schema validation.
```

Comment:
There may be some more rules in here i haven't encoded yet. Either way, this is where we add the custom validation on the raw type.

Reply:
We may want to check for ref integrity here as well. Just make sure that all refs point to a valid target.